src/compact_live.rs

```diff
@@ -0,0 +1,454 @@
+        if (n_compact as f64  * (1f64 + Self::NOISE_RATIO)) > uids.len() as f64 {
+            return Err(Error::CryptoError(format!(
+                "Number of Entry Table UIDs to compact ({n_compact}) should not be greater than {}",
+                uids.len() as f64 / (1f64 + Self::NOISE_RATIO)
+            )));
+        }
+
+        // Upper bounds are used for capacity in order to avoid reallocations:
+        // - the number of noise UIDs is lower than `n_compact`;
+        let mut noise = HashSet::with_capacity(n_compact);
+        // - the total number of UIDs (target set + noise set) is lower than `2*n_compact`.
+        let mut noisy_uids_to_compact = Vec::with_capacity(2 * n_compact);
+
+        // Needed because `uids` is moved in the loop condition.
+        let n_uids = uids.len();
+        let n_noise_candidates = n_uids - n_compact;
+        let n_noise = (Self::NOISE_RATIO * n_compact as f64) as usize;
+
+        for uid in uids {
+            let tmp = rng.next_u32() as usize;
+            if tmp % n_uids < n_compact {
+                // Randomly select ~ `n_compact` UIDs.
+                noisy_uids_to_compact.push(uid);
+            } else if tmp % n_noise_candidates < n_noise {
+                // Randomly select ~ `NOISE_RATIO * n_compact` UIDs.
+                noise.insert(uid);
+            }
+        }
+
+        Ok((noisy_uids_to_compact, noise))
+    }
+
+    /// Fetch all useful information for the compact from the Chain Table:
+    /// - the map of Entry Table UIDs to Chain Table UIDs
+    /// - the map of Entry Table UIDs to indexed values
+    ///
+    /// # Paramters
+    ///
+    /// - `k_value`                 : DEM key used to decrypt the Entry Table
+    /// - `encrypted_entry_table`   : encrypted Entry Table
+    async fn fetch_chain_data(
+        &self,
+        k_value: &DemScheme::Key,
+        encrypted_entry_table: &EncryptedTable<UID_LENGTH>,
+    ) -> Result<
+        ChainData<UID_LENGTH>,
+        Error<CustomError>,
+    > {
+        let entry_table: EntryTable<UID_LENGTH, KWI_LENGTH> =
+            EntryTable::decrypt::<DEM_KEY_LENGTH, DemScheme>(k_value, encrypted_entry_table)?;
+
+        let mut chains = ChainData::with_capacity(entry_table.len());
+
+        // Unchain all Entry Table UIDs.
+        let kwi_chain_table_uids = entry_table.unchain::<
+            CHAIN_TABLE_WIDTH,
+            BLOCK_LENGTH,
+            KMAC_KEY_LENGTH,
+            DEM_KEY_LENGTH,
+            KmacKey,
+            DemScheme
+        >(entry_table.keys(), usize::MAX);
+
+        // Associate Entry Table UIDs to Chain Table UIDs.
+        for (uid, v) in  entry_table.iter() {
+            chains.chain_uids.insert(
+                uid.clone(),
+                kwi_chain_table_uids
+                .get(&v.kwi)
+                .ok_or(Error::<CustomError>::CryptoError(format!(
+                    "no matching Kwi in `kwi_chain_table_uids` ({:?})",
+                    v.kwi
+                )))?.clone(),
```

Comment:
Since `kwi_chain_table_uids` is only used once (and I think `v.kwi` is unique?) you may be able to call `.remove()` instead of `.get` which take ownership of the value and avoid the `.clone()` at the end.

Reply:
You need to call 
```
        let chain_values = self
            .batch_fetch_chains(&kwi_chain_table_uids)
            .await?
```
before by doing it since the `kwi_chain_table_uids` will be empty after the first loop…

Reply:
done in 021cf5b